src/lib/x509/x509_ext.h

```diff
@@ -500,6 +500,65 @@ class OCSP_NoCheck final : public Certificate_Extension {
       void decode_inner(const std::vector<uint8_t>&) override;
 };
 
+/**
+* TNAuthList extension
+*
+* RFC8226 Secure Telephone Identity Credentials
+*   https://www.rfc-editor.org/rfc/rfc8226#section-9
+*/
+class BOTAN_PUBLIC_API(3, 5) TNAuthList final : public Certificate_Extension {
+   public:
+      class Entry final : public ASN1_Object {
+         public:
+            enum Type { TN_ServiceProviderCode = 0, TN_TelephoneNumberRange, TN_TelephoneNumber };
+
+            struct TelephoneNumberRange {
+                  ASN1_String start;  //TelephoneNumber (IA5String)
+                  size_t count;       //2..MAX
+            };
+
+            using RangeContainer = std::vector<TelephoneNumberRange>;
+            using DataContainer = std::variant<ASN1_String, RangeContainer>;
+
+            void encode_into(DER_Encoder&) const override;
+            void decode_from(class BER_Decoder& from) override;
+
+            Type type() const { return m_type; }
+
+            const std::string& service_provider_code() const { return std::get<ASN1_String>(m_data).value(); }
+
+            const RangeContainer& telephone_number_range() const { return std::get<RangeContainer>(m_data); }
+
+            const std::string& telephone_number() const { return std::get<ASN1_String>(m_data).value(); }
```

Comment:
Yay! `std::variant<>`. Thanks a lot!

One small user experience thing: If someone calls the wrong accessor, they'll get a weird `std::bad_variant_access` exception that they may or may not understand.

Let's make those a bit more friendly:

```suggestion
            const std::string& service_provider_code() const {
                BOTAN_STATE_CHECK(type() == Type::TN_ServiceProviderCode);
                return std::get<ASN1_String>(m_data).value();
            }

            const RangeContainer& telephone_number_range() const {
                BOTAN_STATE_CHECK(type() == Type::TN_TelephoneNumberRange);
                return std::get<RangeContainer>(m_data);
            }

            const std::string& telephone_number() const {
                BOTAN_STATE_CHECK(type() == Type::TN_TelephoneNumber);    
                return std::get<ASN1_String>(m_data).value();
            }
```

... that'll produce `Botan::Invalid_State` with a reasonable error message when we're not in the correct state for the respective accessor.

Reply:
Not only that - since the variant shares a type (`ASN1_String`) for two of the possible data types, currently if `type() == TN_TelephoneNumber` then `service_provider_code()` works and returns the wrong thing.